File: 2024/typescript/day04.ts
import { type DayModule, type SolutionFactory } from './types'
import * as utils from './utils'

const XMAS = 'XMAS'

type CounterFn = (lines: string[], x: number, y: number) => number

function count_XMAS(lines: string[], x: number, y: number): number {
    if (lines[y][x] !== XMAS[0]) {
        return 0
    }

    let directions: [number, number][] = [
        [1, 0],
        [-1, 0],
        [0, 1],
        [0, -1],
        [1, 1],
        [-1, -1],
        [1, -1],
        [-1, 1],
    ]

    return Array.from(XMAS.slice(1), (c, i) => [c, i + 1] as const).reduce(
        (directions, [c, i]) =>
            directions.filter(
                ([dx, dy]) => lines[y + dy * i]?.[x + dx * i] === c,
            ),
        directions,
    ).length
}

function count_MAS_cross(lines: string[], x: number, y: number): number {
    const diag1 = () => lines[y + 1]?.[x + 1] + lines[y - 1]?.[x - 1]
    const diag2 = () => lines[y + 1]?.[x - 1] + lines[y - 1]?.[x + 1]
    const isMS = (s: string) => s === 'MS' || s === 'SM'

    return lines[y][x] === 'A' && isMS(diag1()) && isMS(diag2()) ? 1 : 0
}

function count(lines: string[], counter: CounterFn): number {
    return lines.reduce(
        (sum, _, y) =>
            _.split('').reduce((sum, _, x) => sum + counter(lines, x, y), sum),
        0,
    )
}

export const part1: Part = input => () => count(input, count_XMAS)

export const part2: Part = input => () => count(input, count_MAS_cross)

export const day = import.meta.file.match(/day(\d+)/)![1]
export const input = await utils.readInputLines(day)
part1.solution = 2517
part2.solution = 1960

export const main = import.meta.main
if (main) {
    const module: Module = await import(import.meta.path)

    const testInput = utils.as_lines([
        `
MMMSXXMASM
MSAMXMSMSA
AMXSXMAAMM
MSAMASMSMX
XMASAMXAMM
XXAMMXXAMA
SMSMSASXSS
SAXAMASAAA
MAMMMXMMMM
MXMXAXMASX
`,
    ])

    await utils.tests(
        () =>
            utils.test_all(
                ['Test part 1', 18, part1(testInput[0])],
                ['Test part 2', 9, part2(testInput[0])],
            ),
        () => utils.test_day(module),
    )
}

type In = typeof input
type Out = number
type Module = DayModule<Out, In>
type Part = SolutionFactory<Out, In>

File: 2024/zig/src/day04.zig
const std = @import("std");

const aoc = @import("aoc_utils");

fn part1(input: []const u8, gpa: std.mem.Allocator) !u32 {
    const gridBuffer = try gpa.dupe(u8, input);
    defer gpa.free(gridBuffer);
    const grid = aoc.Grid.init(gridBuffer);
    var result: u32 = 0;
    var p = aoc.Pos{ .x = 0, .y = 0 };
    while (p.y < grid.height) : (p.y += 1) {
        p.x = 0;
        while (p.x < grid.width) : (p.x += 1) {
            if (grid.at(p) != 'X') continue;
            inline for (@typeInfo(aoc.Direction).@"enum".fields) |field| {
                const dir = @field(aoc.Direction, field.name);
                if (grid.subarray(3, p, dir, 1)) |word| {
                    if (std.mem.eql(u8, &word, "MAS")) result += 1;
                }
            }
        }
    }
    return result;
}

fn part2(input: []const u8, gpa: std.mem.Allocator) !u32 {
    const gridBuffer = try gpa.dupe(u8, input);
    defer gpa.free(gridBuffer);
    const grid = aoc.Grid.init(gridBuffer);
    var result: u32 = 0;
    var p = aoc.Pos{ .x = 1, .y = 1 };
    while (p.y < grid.height - 1) : (p.y += 1) {
        p.x = 1;
        while (p.x < grid.width - 1) : (p.x += 1) {
            if (grid.at(p) != 'A') {
                continue;
            }
            const corners = [_]u8{
                grid.at(aoc.Direction.north_west.next(p)),
                grid.at(aoc.Direction.north_east.next(p)),
                grid.at(aoc.Direction.south_west.next(p)),
                grid.at(aoc.Direction.south_east.next(p)),
            };
            if (std.mem.eql(u8, &corners, "MMSS") or
                std.mem.eql(u8, &corners, "SSMM") or
                std.mem.eql(u8, &corners, "MSMS") or
                std.mem.eql(u8, &corners, "SMSM")) result += 1;
        }
    }
    return result;
}

const Day = aoc.DayInfo("04", u32, u32, 2517, 1960, &.{
    .{
        .expected1 = 12,
        .expected2 = null,
        .input =
        \\XMAS.SAMX
        \\MM.....MM
        \\A.A...A.A
        \\S..S.S..S
        \\.........
        \\S..S.S..S
        \\A.A...A.A
        \\MM.....MM
        \\XMAS.SAMX
        ,
    },
    .{
        .expected1 = 3,
        .expected2 = null,
        .input =
        \\XMAS
        \\MM..
        \\A.A.
        \\S..S
        ,
    },
    .{
        .expected1 = 3,
        .expected2 = null,
        .input =
        \\XMAS
        \\MM..
        \\A.A.
        \\S..S
        ,
    },
    .{
        .expected1 = 3,
        .expected2 = null,
        .input =
        \\S..S
        \\A.A.
        \\MM..
        \\XMAS
        ,
    },
    .{
        .expected1 = 3,
        .expected2 = null,
        .input =
        \\SAMX
        \\..MM
        \\.A.A
        \\S..S
        ,
    },
    .{
        .expected1 = 3,
        .expected2 = null,
        .input =
        \\S..S
        \\.A.A
        \\..MM
        \\SAMX
        ,
    },
    .{
        .expected1 = 18,
        .expected2 = 9,
        .input =
        \\MMMSXXMASM
        \\MSAMXMSMSA
        \\AMXSXMAAMM
        \\MSAMASMSMX
        \\XMASAMXAMM
        \\XXAMMXXAMA
        \\SMSMSASXSS
        \\SAXAMASAAA
        \\MAMMMXMMMM
        \\MXMXAXMASX
        ,
    },
});

test "samples 1" {
    try Day.testPart1Samples(part1);
}
test "samples 2" {
    try Day.testPart2Samples(part2);
}
test "part 1" {
    try Day.testPart1(part1);
}
test "part 2" {
    try Day.testPart2(part2);
}
